### PR TITLE
Octopus and SES7 deployment with sesboot

### PR DESCRIPTION
**Pull request by rjfd**
Besides the implementation of octopus deployment using sesboot, this PR also changes the way the deployment bash scripts are included in the Vagrantfile.
Instead of inline the bash script in the Vagrantfile (which led to several problems when the bash script was escaping special characters in sed expressions), the bash scripts are now stored in a file and the file is uploaded and used automatically by Vagrant.

To test this PR and deploy an Octopus cluster using sesboot/ceph-daemon/ssh-orchestrator you can issue the following command:

```
sesdev create octopus --repo https://download.opensuse.org/repositories/home:/rjdias:/branches:/openSUSE:/Factory/openSUSE_Leap_15.2/ octopus-test
```

The additional repo is used to provide a version of podman that works well when running ceph-daemon commands from the salt-minion process.
By default, sesdev will install sesboot and ses-formula from RPM packages available in https://build.opensuse.org/project/show/home:rjdias:sesboot
But sesdev can also install sesboot and ses-formula from source using some CLI options added in this PR.

sesdev will not change the default ceph container image used by ceph-daemon and thus will use the docker.io/ceph/daemon-base:latest-master-devel image. You can however explicitly specify the path for another container image using the `--ceph-container-image` CLI option.

### CODE REVIEW

**Comment by rjfd** [comment]
Very soon `sesboot` will be renamed `ceph-bootstrap`, and `ses-formula` renamed to `ceph-salt-formula`. 

The new projects will reside in:
  * https://github.com/SUSE/ceph-bootstrap
  * https://github.com/SUSE/ceph-salt-formula

I'll change this PR accordingly, when the above change takes effect.

File: seslib/templates/provision.sh.j2 [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+{% for _node in nodes %}
+echo "{{ _node.public_address }} {{ _node.fqdn }} {{ _node.name }}" >> /etc/hosts
+{% endfor %}
+
+{% for os_repo_name, os_repo_url in os_base_repos %}
+zypper ar {{ os_repo_url }} {{ os_repo_name }}
+{% endfor %}
+
+zypper ar {{ version_repo }} {{ version }}-repo
```
Comment (smithfarm) [inline seslib/templates/provision.sh.j2]: Based on past experience, I suggest to do something like:

```
ZYPPER_ADDREPO="zypper --non-interactive addrepo --refresh --no-gpgcheck"
$ZYPPER_ADDREPO {{ os_repo_url }} {{ os_repo_name }}
```

File: seslib/templates/provision.sh.j2 [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+zypper --gpg-auto-import-keys ref
```
Comment (smithfarm) [inline seslib/templates/provision.sh.j2]: Suggested improvement:

```
ZYPPER_REFRESH="zypper --non-interactive --no-gpg-checks refresh"
$ZYPPER_REFRESH
```

File: seslib/templates/provision.sh.j2 [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+{% if version == 'ses5' %}
+zypper -n install vim git-core iputils jq make iptables patch
```
Comment (smithfarm) [inline seslib/templates/provision.sh.j2]:
```
ZYPPER_INSTALL="zypper --non-interactive --gpg-auto-import-keys install --no-recommends"
$ZYPPER_INSTALL vim git-core iputils jq make iptables patch
```

File: seslib/templates/provision.sh.j2 [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+{% else %}
+zypper -n install vim git iputils hostname jq make iptables patch
+{% endif %}
+
+{% for repo in node.repos %}
+zypper ar {{ repo.url }} {{ repo.name }}
```
Comment (smithfarm) [inline seslib/templates/provision.sh.j2]:
```
$ZYPPER_ADDREPO {{ repo.url }} {{ repo.name }}
```

File: seslib/templates/provision.sh.j2 [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+{% if repo_priority and repo.priority %}
+zypper mr -p {{ repo.priority }} {{ repo.name }}
+{% endif %}
+{% endfor %}
+{% if node.repos|length > 0 %}
+zypper --gpg-auto-import-keys ref
```
Comment (smithfarm) [inline seslib/templates/provision.sh.j2]:
```
$ZYPPER_REFRESH
```

File: seslib/templates/provision.sh.j2 [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+cat /home/vagrant/.ssh/id_rsa.pub >> /home/vagrant/.ssh/authorized_keys
+[ ! -e "/root/.ssh" ] && mkdir /root/.ssh
+chmod 600 /home/vagrant/.ssh/id_rsa
+cp /home/vagrant/.ssh/id_rsa* /root/.ssh/
+cat /root/.ssh/id_rsa.pub >> /root/.ssh/authorized_keys
+hostnamectl set-hostname {{ node.name }}
+
+{% if not suma %}
+zypper -n install salt-minion
```
Comment (smithfarm) [inline seslib/templates/provision.sh.j2]:
```
$ZYPPER_INSTALL salt-minion
```

File: seslib/templates/sesboot/sesboot_deployment.sh.j2 [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+{% if sesboot_git_repo %}
+cd /root
+git clone {{ sesboot_git_repo }}
+cd sesboot
+zypper -n in autoconf gcc python3-devel python3-pip
+git checkout {{ sesboot_git_branch }}
+pip install .
+{% else %}
+{% if os == 'leap-15.2' %}
+zypper ar https://download.opensuse.org/repositories/home:/rjdias:/sesboot/openSUSE_Leap_15.2/ sesboot_repo
```
Comment (smithfarm) [inline seslib/templates/sesboot/sesboot_deployment.sh.j2]:
```
ZYPPER_ADDREPO="zypper --non-interactive addrepo --refresh --no-gpgcheck"
$ZYPPER_ADDREPO {{ os_repo_url }} {{ os_repo_name }}
```

File: seslib/templates/sesboot/sesboot_deployment.sh.j2 [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+{% endif %}
+{% if os == 'tumbleweed' %}
+zypper ar https://download.opensuse.org/repositories/home:/rjdias:/sesboot/openSUSE_Tumbleweed/ sesboot_repo
```
Comment (smithfarm) [inline seslib/templates/sesboot/sesboot_deployment.sh.j2]:
```
$ZYPPER_ADDREPO https://download.opensuse.org/repositories/home:/rjdias:/sesboot/openSUSE_Tumbleweed/ sesboot_repo
```

File: seslib/templates/sesboot/sesboot_deployment.sh.j2 [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+{% endif %}
+
+zypper --gpg-auto-import-keys ref
```
Comment (smithfarm) [inline seslib/templates/sesboot/sesboot_deployment.sh.j2]:
```
ZYPPER_REFRESH="zypper --non-interactive --no-gpg-checks refresh"
$ZYPPER_REFRESH
```

File: seslib/templates/sesboot/sesboot_deployment.sh.j2 [hunk 10]
```diff
@@ -0,0 +1,95 @@
+
+
+{% if sesboot_git_repo %}
+cd /root
+git clone {{ sesboot_git_repo }}
+cd sesboot
+zypper -n in autoconf gcc python3-devel python3-pip
```
Comment (smithfarm) [inline seslib/templates/sesboot/sesboot_deployment.sh.j2]:
```
ZYPPER_INSTALL="zypper --non-interactive --gpg-auto-import-keys install --no-recommends"
$ZYPPER_INSTALL autoconf gcc python3-devel python3-pip
```

File: seslib/templates/sesboot/sesboot_deployment.sh.j2 [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+{% if os == 'leap-15.2' %}
+zypper ar https://download.opensuse.org/repositories/home:/rjdias:/sesboot/openSUSE_Leap_15.2/ sesboot_repo
+{% endif %}
+{% if os == 'tumbleweed' %}
+zypper ar https://download.opensuse.org/repositories/home:/rjdias:/sesboot/openSUSE_Tumbleweed/ sesboot_repo
+{% endif %}
+
+zypper --gpg-auto-import-keys ref
+# ses-formula is installed automatically as a dependency of sesboot
+zypper -n in sesboot
```
Comment (smithfarm) [inline seslib/templates/sesboot/sesboot_deployment.sh.j2]:
```
$ZYPPER_INSTALL sesboot
```

**Comment by rjfd** [comment]
I just updated the PR with the change from sesboot to ceph-bootstrap. It also now supports deploying version `ses7`.

**Comment by smithfarm** [comment]
@rjfd Is it expected that `--single-node` doesn't work with `octopus`/`ses7` ?

File: setup.cfg [hunk 12]
```diff
@@ -35,6 +35,7 @@ packages =
 seslib =
     templates/*.j2
     templates/deepsea/*.j2
+    templates/sesboot/*.j2
```
Comment (smithfarm) [inline setup.cfg]: Does `sesboot` need to be changed to `ceph-bootstrap` here?

Comment (rjfd) [reply setup.cfg]: yes, it needs. Good catch!

**Comment by rjfd** [comment]
> @rjfd Is it expected that `--single-node` doesn't work with `octopus`/`ses7` ?

No, it's not. It should work, if it doesn't I need to fix it.

**Comment by smithfarm** [comment]
@rjfd See http://pastebin.nue.suse.com/23212/src for the full typescript showing how `--single-node` failed in my environment.

**Comment by rjfd** [comment]
> @rjfd See http://pastebin.nue.suse.com/23212/src for the full typescript showing how `--single-node` failed in my environment.

I pushed a commit with a fix.

File: setup.cfg (hunk 12 above)
Comment (rjfd) [reply setup.cfg]: done

File: seslib/templates/provision.sh.j2 (hunk 1 above)
Comment (rjfd) [reply seslib/templates/provision.sh.j2]: Can we do these changes in another PR?

Comment (smithfarm) [reply seslib/templates/provision.sh.j2]: Sure.

These suggestions were not meant to block this PR. I tested it, and it works even without these changes.

**Pull request merged**